src/Infrastructure/KafkaJanitor.RestApi/Features/Access/Application/AccessService.cs

```diff
@@ -0,0 +1,108 @@
+using System.Threading.Tasks;
+using KafkaJanitor.RestApi.Features.AccessControlLists.Infrastructure;
+using KafkaJanitor.RestApi.Features.ApiKeys;
+using KafkaJanitor.RestApi.Features.ServiceAccounts.Infrastructure;
+using KafkaJanitor.RestApi.Features.Topics.Domain.Models;
+using KafkaJanitor.RestApi.Features.Vault;
+using KafkaJanitor.RestApi.Features.Vault.Model;
+using Tika.RestClient.Features.ServiceAccounts.Models;
+
+namespace KafkaJanitor.RestApi.Features.Access.Application
+{
+    public class AccessService : IAccessService
+    {
+        private readonly IAccessControlListClient _accessControlListService;
+        private readonly IServiceAccountClient _serviceAccountClient;
+        private readonly IApiKeyClient _apiKeyClient;
+        private readonly IVault _vault;
+
+        public AccessService(
+            IAccessControlListClient accessControlListService,
+            IServiceAccountClient serviceAccountClient,
+            IApiKeyClient apiKeyClient,
+            IVault vault
+        )
+        {
+            _accessControlListService = accessControlListService;
+            _serviceAccountClient = serviceAccountClient;
+            _apiKeyClient = apiKeyClient;
+            _vault = vault;
+        }
+
+
+        public async Task ProvideAccess(
+            Capability capability,
+            string topicPrefix
+        )
+        {
+            ServiceAccount serviceAccount;
+            if (!await ServiceAccountExists(capability))
+            {
+                serviceAccount = await _serviceAccountClient.CreateServiceAccount(capability);
+            }
+            else
+            {
+                serviceAccount = await _serviceAccountClient.GetServiceAccount(capability);
+            }
+
+            if (!await ExpectedAmountOfAclsAreInPlace(serviceAccount.Id))
+            {
+                await _accessControlListService.CreateAclsForServiceAccount(serviceAccount.Id, topicPrefix);
+            }
+
+            if (!await AtLestOneApiKeyExists(serviceAccount.Id))
+            {
+                await CreateAndStoreApiKeyPair(
+                    capability, 
+                    serviceAccount
+                );
+            }
+        }
+
+        public async Task CreateAndStoreApiKeyPair(
+            Capability cap, 
+            ServiceAccount serviceAccount
+        )
+        {
+            var apiKeyPair = await _apiKeyClient.CreateApiKeyPair(serviceAccount);
+
+            await _vault.AddApiCredentials(
+                cap,
+                new ApiCredentials
+                {
+                    Key = apiKeyPair.Key,
+                    Secret = apiKeyPair.Secret
+                }
+            );
+        }
+        
+        public async Task<bool> ServiceAccountExists(Capability capability)
+        {
+            try
+            {
+                var sa = await _serviceAccountClient.GetServiceAccount(capability);
+            }
+            catch (InvalidOperationException ex)
```

Comment:
This smells fishy

Does the `InvalidOperationException`raise on `results.First(sa => sa.Name == $"{capability.Name}_sa")` == null?

consider a ServiceAccountDoesNotExist exception instead.